### ISSUE TITLE
Mempool admission doesn't handle unlocks correctly

### ISSUE THREAD

**Issue by kantai** [stacks-network/stacks-blockchain#3623]
This is not a consensus-critical bug, but it is something that can cause some user headaches.

The current mempool admission logic doesn't correctly handle unlocks and could tell people who have PoX-locked all of their account funds and try to transfer (or do a contract-call with a very high fee) after an unlock that they don't have enough funds for their transaction.

The consensus codebase performs "lazy unlocks" (this is true of both the PoX-2 unlock and the old way that PoX unlocked at the end of a cycle): the code stores the height at which an unlock occurs and only actually alters the account entry to "unlocked" when the account changes (via credit, debit, or applying a new lock). The mempool admission logic doesn't check the lazy unlock correctly: it will think an account is still locked up until the unlock actually materializes via one of those changes. I'm pretty sure this bug has existed since 2.0 launch (and would have come up if someone tried to transfer immediately after an unlock).

The patch for this is fairly straightforward. Fundamentally, it looks something like this:

```diff
--- a/src/chainstate/stacks/db/blocks.rs
+++ b/src/chainstate/stacks/db/blocks.rs
@@ -6975,6 +6975,7 @@ impl StacksChainState {
     /// unconfirmed microblock stream trailing off of it.
     pub fn will_admit_mempool_tx(
         &mut self,
+        sortdb: &SortitionDB,
         current_consensus_hash: &ConsensusHash,
         current_block: &BlockHeaderHash,
         tx: &StacksTransaction,
@@ -7019,7 +7020,7 @@ impl StacksChainState {
 
         let current_tip =
             StacksChainState::get_parent_index_block(current_consensus_hash, current_block);
-        let res = match self.with_read_only_clarity_tx(&NULL_BURN_STATE_DB, &current_tip, |conn| {
+        let res = match self.with_read_only_clarity_tx(&sortdb.index_conn(), &current_tip, |conn| {
             StacksChainState::can_include_tx(conn, &conf, has_microblock_pubk, tx, tx_size)
         }) {
             Some(r) => r,
@@ -7039,7 +7040,7 @@ impl StacksChainState {
                 {
                     debug!("Transaction {} is unminable in the confirmed chain tip due to nonce {} != {}; trying the unconfirmed chain tip",
                            &tx.txid(), mismatch_error.expected, mismatch_error.actual);
-                    self.with_read_only_unconfirmed_clarity_tx(&NULL_BURN_STATE_DB, |conn| {
+                    self.with_read_only_unconfirmed_clarity_tx(&sortdb.index_conn(), |conn| {
                         StacksChainState::can_include_tx(
                             conn,
                             &conf,
```

This should be hotfixed. However, note that actually remedying the user pain here will require that a sufficient number of miners and nodes in the network upgrade to the hotfixed version. In the meantime, user workarounds are possible.

## User workarounds

Users can work around this issue before nodes upgrade by sending any transaction that performs a credit or debit action on their account. For example:

1. Send a small amount of STX from a different address to the account they want to unlock.
2. Send a very small transfer from the account they want to unlock to another account.
3. Perform a contract-call from the account they want to unlock.

**Comment by diwakergupta** [comment]
Is there a way to qualify the impact -- does this affect all accounts and unlocks, or only some? If latter, is there a way to characterize which accounts will run into this?

**Comment by kantai** [comment]
> Is there a way to qualify the impact -- does this affect all accounts and unlocks, or only some? If latter, is there a way to characterize which accounts will run into this?

It affects all accounts that unlocked some amount at the PoX auto-unlock (so any account that was locked in PoX). Not all of those accounts will "notice" this issue, though, because if your previously unlocked amount has enough funds to cover a transaction fee, you can do a contract-call with your account which would eliminate the issue.

**Comment by 314159265359879** [comment]
I like that there are clear workarounds, but indeed causing some headaches as @kantai pointed out. 

Do I understand correctly this is already fixed? I believe the fix is in the 2.1.0.0.2 pre-release, when will that be released?

**Comment by kantai** [comment]
Ah -- the release on github was still just tagged as pre-release. It is now marked as the latest release.
